cabal-install/tests/IntegrationTests2.hs

```diff
@@ -94,20 +98,42 @@ testExceptionInBuildStep = do
     plan  <- planProject testdir config
     plan' <- executePlan plan
     (_pkga1, failure) <- expectPackageFailed plan' pkgidA1
-    case failure of
-      BuildFailed _str -> return ()
-      _ -> assertFailure $ "expected BuildFailed, got " ++ show failure 
+    expectBuildFailed failure
   where
     testdir = "exception/build"
     config  = mempty
     pkgidA1 = PackageIdentifier (PackageName "a") (Version [1] [])
 
+-- | See <https://github.com/haskell/cabal/issues/3324>
+--
+testRegressionIssue3324 :: Assertion
+testRegressionIssue3324 = do
+    -- expected failure first time due to missing dep
+    plan1 <- executePlan =<< planProject testdir config
+    (_pkgq, failure) <- expectPackageFailed plan1 pkgidQ
+    expectBuildFailed failure
+
+    -- add the missing dep, now it should work
+    appendFile (basedir </> testdir </> "q" </> "q.cabal")
```

Comment:
I think you need to add a `finally` block restoring the `q.cabal` file to its original state, otherwise the test suite is not idempotent.